### PR TITLE
Ear 1100 move qCode validation client side

### PR DESCRIPTION
**Pull request by zimboflyman**
### What is the context of this PR?
This PR covers 2 Tickets:
EAR-1100
EAR-977

Devs agreed that it is better to refactor the QCode validation page due to complexities with the existing implementation. 
Remove all AJV refs to qCodes
Remove qcodeErrorTotals from resolvers
Move all qCode validation to clientside
Move answer state (Previously getAllAnswers.graphql) up to QuestionnaireDesignPage
Use React.Context to pass list of answers AND Duplicate QCodes down to qCode page.
Update Queries in QuestionnaireDesignPage so as update the context when Qcodes are added in qCode page - This will allow us to handle red error dot in mainNav
Red Dot to show in Main NAv when:
1: an empty QCode
2: Duplicate QCodes

### How to review
Make a list of all question types
You should see red dot on qCode icon of main menu
Add qCodes to all answers
Red dot should disappear or reflect missing qcodes / duplicates

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process

### CODE REVIEW

File: eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.js [hunk 1]
```diff
@@ -85,119 +88,128 @@ export const UnwrappedMainNavigation = props => {
 
   const previewUrl = `${config.REACT_APP_LAUNCH_URL}/${
     (questionnaire || {}).id
-    }`;
+  }`;
+
+  const emptyQCode =
+    flattenedAnswers &&
+    find(flattenedAnswers, obj => obj.qCode === "" || obj.qCode === null);
```
Comment (tbaggaley) [inline eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.js:95]:
```suggestion
const emptyQCode = flattenedAnswers?.find(({type, qCode}) => type !== "Checkbox" && !qCode);
```

This fixes issue where Checkbox qCodes were coming up as missing (for the parent question) on imported LCS.

**Review by tbaggaley — CHANGES_REQUESTED** [review]
Hey Shane - looking good so far, haven't finished reviewing yet but there's a couple small changes here and one to fix the red dot coming up when parent checkbox answers have empty qCodes (as seen when you import LCS)

File: eq-author/src/App/qcodes/QCodesTable/index.js [hunk 2]
```diff
@@ -333,36 +223,22 @@ const Row = memo(props => {
...
+      return (
+        <Row
+          key={`${item.id}-${index}`}
+          {...item}
+          error={duplicates[item.qCode] > 1}
+          noValQCodeError={noValQCodeError}
+        />
+      );
+    });
+  }
```
Comment (tbaggaley) [inline eq-author/src/App/qcodes/QCodesTable/index.js:241]:
```suggestion
  if(answers) {
    return answers.map((item, index) =>
      <Row
        key={`${item.id}-${index}`}
        {...item}
        error={duplicates[item.qCode] > 1}
        noValQCodeError={!item.qCode}
      />);
    }
```

File: eq-author-api/src/validation/index.js [hunk 3]
```diff
@@ -27,17 +27,7 @@ const formatErrorMessage = (error, questionnaire) => {
 
   const splitDataPath = dataPath.split("/");
   let field = "";
```
[the remark this answers is not in the document]
Comment (zimboflyman) [reply eq-author-api/src/validation/index.js:29]: haha - makes sense!

**Review by tbaggaley — CHANGES_REQUESTED** [review]

**Review by tbaggaley — CHANGES_REQUESTED** [review]
Could we move `QCodeContext` out into its own file too (like components/QuestionnaireContext)?

It could export a `useQCodeContext()` hook for use in client code elsewhere - like with the questionnaire context one - saves having to directly import the `QCodeContext` / `useContext` on each file

File: eq-author/src/App/qcodes/QCodesTable/index.js (hunk 2 above)
Comment (zimboflyman) [reply eq-author/src/App/qcodes/QCodesTable/index.js:241]: much tidier - ta

Comment (zimboflyman) [reply eq-author/src/App/qcodes/QCodesTable/index.js:241]: much tidier - ta

**Comment by zimboflyman** [comment]
> components/QuestionnaireContext

Good idea Tom - done

File: eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js [hunk 4]
```diff
@@ -3,10 +3,11 @@ import { render, flushPromises } from "tests/utils/rtl";
 import { UnwrappedMainNavigation, publishStatusSubscription } from "./";
 import { MeContext } from "App/MeContext";
 import { act } from "react-dom/test-utils";
-//import { SynchronousPromise } from "synchronous-promise";
+import { QCodeContext } from "App/QuestionnaireDesignPage";
```
Comment (farres1) [inline eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js:6]:
```suggestion
import { QCodeContext } from "components/QCodeContext";
```

File: eq-author/src/App/qcodes/QCodesTable/index.test.js [hunk 5]
```diff
@@ -6,22 +6,9 @@ import UPDATE_OPTION_QCODE from "./graphql/updateOptionMutation.graphql";
 import UPDATE_CONFIRMATION_QCODE from "./graphql/updateConfirmationQCode.graphql";
 
 import { MeContext } from "App/MeContext";
+import { QCodeContext } from "App/QuestionnaireDesignPage";
```
Comment (farres1) [inline eq-author/src/App/qcodes/QCodesTable/index.test.js:9]:
```suggestion
import { QCodeContext } from "components/QCodeContext";
```

**Review by farres1 — CHANGES_REQUESTED** [review]

File: eq-author/src/App/qcodes/QCodesTable/index.test.js (hunk 5 above)
Comment (zimboflyman) [reply eq-author/src/App/qcodes/QCodesTable/index.test.js]: Thanks Sam - I just uploaded this - ta!

File: eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js (hunk 4 above)
Comment (zimboflyman) [reply eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js]: thanks Sam - sorted now!

File: eq-author/src/App/qcodes/QCodesTable/index.js [hunk 6]
```diff
@@ -333,36 +223,20 @@ const Row = memo(props => {
...
-    return (
-      <Row
-        key={`${item.id}-${index}`}
-        {...item}
-        error={duplicates[item.qCode] > 1}
-        noValQCodeError={noValQCodeError}
-      />
-    );
-  });
+  if(answers) {
```
Comment (tjamesmac) [inline eq-author/src/App/qcodes/QCodesTable/index.js:228]:
```suggestion
  if (answers) {
```

File: eq-author/src/App/qcodes/QCodesTable/index.js [hunk 7]
```diff
@@ -375,21 +249,12 @@ Row.propTypes = {
   qCodeCheck: PropTypes.func,
   error: PropTypes.bool,
   nested: PropTypes.bool,
-  noValQCodeError: PropTypes.object, // eslint-disable-line
+  noValQCodeError: PropTypes.bool, // eslint-disable-line
```
Comment (tjamesmac) [inline eq-author/src/App/qcodes/QCodesTable/index.js:252]:
```suggestion
  noValQCodeError: PropTypes.bool,
```

File: eq-author/src/App/qcodes/QCodesTable/index.js [hunk 8]
```diff
@@ -375,21 +249,12 @@ Row.propTypes = {
...
-
-  if (error) {
-    return <Error>Oops! Something went wrong</Error>;
-  }
+export const UnwrappedQCodeTable = () => {
 
-  const { sections } = data.questionnaire;
-  const { answers } = organiseAnswers(sections);
-  const flatten = flattenAnswers(answers);
+  const {flattenedAnswers, duplicates }  = useQCodeContext();
```
Comment (tjamesmac) [inline eq-author/src/App/qcodes/QCodesTable/index.js:257]:
```suggestion
  const { flattenedAnswers, duplicates }  = useQCodeContext();
```

File: eq-author/src/utils/getAllAnswersFlatMap.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
+/* eslint-disable no-loop-func */
+import {
+    RADIO,
+  } from "constants/answer-types";
```
Comment (tjamesmac) [inline eq-author/src/utils/getAllAnswersFlatMap.js:4]:
```suggestion
import { RADIO } from "constants/answer-types";
```

File: eq-author/src/App/qcodes/QCodesTable/index.test.js [hunk 10]
```diff
@@ -34,305 +21,278 @@ const dummyQcodes = {
...
+        id: 'option-cb-2',
+        label: 'Or',
+        mutuallyExclusive: true,
+        description: null,
+        additionalAnswer: null,
+        qCode: '20',
+        type: 'MutuallyExclusiveOption',
+        option: true 
+      } 
+  ];
```
Comment (tjamesmac) [inline eq-author/src/App/qcodes/QCodesTable/index.test.js:290]:
```suggestion
    flattenedAnswers = [
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        id: "ans-p1-1",
        description: "",
        guidance: "",
        label: "num1",
        qCode: "123",
        secondaryQCode: "1",
        type: "Number",
        questionPageId: "qp-1",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "ans-p1-2",
        description: "",
        guidance: "",
        label: "curr1",
        qCode: "123",
        secondaryQCode: "2",
        type: "Currency",
        questionPageId: "qp-1",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "ans-p1-3",
        description: "",
        guidance: "",
        label: "Un1",
        qCode: "1",
        secondaryQCode: "3",
        type: "Unit",
        questionPageId: "qp-1",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "ans-p1-4",
        description: "",
        guidance: "",
        label: "Per1",
        qCode: "2",
        secondaryQCode: "4",
        type: "Percentage",
        questionPageId: "qp-1",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "ans-p1-5",
        description: "",
        guidance: "",
        label: "Dur1",
        qCode: "3",
        secondaryQCode: "5",
        type: "Duration",
        questionPageId: "qp-1",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "ans-p1-6",
        description: "",
        guidance: "",
        label: "Num2",
        qCode: "4",
        secondaryQCode: "5",
        type: "Number",
        questionPageId: "qp-1",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "ans-p7-1",
        description: "",
        guidance: "",
        label: "",
        qCode: "5",
        type: "Checkbox",
        questionPageId: "29ceee38-5ba4-4f43-84ae-0162c5b175f8",
        options: [[Object], [Object]],
        mutuallyExclusiveOption: {
          id: "cb-3",
          label: "Embedded checkbox Or",
          mutuallyExclusive: true,
          description: null,
          additionalAnswer: null,
          qCode: "29",
        },
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "cb-1",
        label: "Embedded checkbox Either",
        description: null,
        additionalAnswer: null,
        qCode: "27",
        type: "CheckboxOption",
        option: true,
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "cb-2",
        label: "Either 2",
        description: null,
        additionalAnswer: null,
        qCode: "28",
        type: "CheckboxOption",
        option: true,
      },
      {
        title: "<p>Questions 1</p>",
        alias: undefined,
        nested: true,
        id: "cb-3",
        label: "Embedded checkbox Or",
        mutuallyExclusive: true,
        description: null,
        additionalAnswer: null,
        qCode: "29",
        type: "MutuallyExclusiveOption",
        option: true,
      },
      {
        title: "<p>Questions 2</p>",
        alias: undefined,
        id: "ans-p2-1",
        description: "",
        guidance: "",
        label: "Da1",
        qCode: "",
        secondaryQCode: "6",
        type: "Date",
        questionPageId: "qp-2",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 3</p>",
        alias: undefined,
        id: "ans-p3-1",
        description: "",
        guidance: "",
        label: "To",
        secondaryLabel: "From",
        qCode: "6",
        secondaryQCode: "99",
        type: "DateRange",
        questionPageId: "qp-3",
      },
      {
        title: "<p>Questions 3</p>",
        alias: undefined,
        nested: true,
        id: "ans-p3-1",
        label: "From",
        qCode: "99",
        type: "DateRange",
        validationErrorInfo: undefined,
        secondary: true,
      },
      {
        title: "<p>Questions 4</p>",
        alias: undefined,
        id: "ans-p4-1",
        description: "",
        guidance: "",
        label: "TF",
        qCode: "",
        secondaryQCode: "7",
        type: "TextField",
        questionPageId: "qp-4",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 4</p>",
        alias: undefined,
        nested: true,
        id: "ans-p4-2",
        description: "",
        guidance: "",
        label: "TA",
        qCode: "",
        secondaryQCode: "8",
        type: "TextArea",
        questionPageId: "qp-4",
        secondaryLabel: null,
      },
      {
        title: "<p>Questions 5</p>",
        alias: undefined,
        id: "conf-q-1",
        qCode: "10",
        type: "QuestionConfirmation",
        validationErrorInfo: undefined,
      },
      {
        title: "<p>Questions 7</p>",
        alias: null,
        id: "ans-p6-1",
        description: "",
        guidance: "",
        label: "",
        qCode: "",
        type: "Radio",
        questionPageId: "qp-6",
        options: [[Object], [Object]],
      },
      {
        title: "<p>Questions 8</p>",
        alias: undefined,
        id: "ans-p7-1",
        description: "",
        guidance: "",
        label: "",
        qCode: "1238",
        type: "Checkbox",
        questionPageId: "29ceee38-5ba4-4f43-84ae-0162c5b175f8",
        options: [[Object], [Object]],
        mutuallyExclusiveOption: {
          id: "option-cb-2",
          label: "Or",
          mutuallyExclusive: true,
          description: null,
          additionalAnswer: null,
          qCode: "20",
        },
      },
      {
        title: "<p>Questions 8</p>",
        alias: undefined,
        nested: true,
        id: "option-cb-1",
        label: "Either",
        description: null,
        additionalAnswer: null,
        qCode: "9",
        type: "CheckboxOption",
        option: true,
      },
      {
        title: "<p>Questions 8</p>",
        alias: undefined,
        nested: true,
        id: "option-cb-3",
        label: "Either 2",
        description: null,
        additionalAnswer: null,
        qCode: "chk",
        type: "CheckboxOption",
        option: true,
      },
      {
        title: "<p>Questions 8</p>",
        alias: undefined,
        nested: true,
        id: "option-cb-2",
        label: "Or",
        mutuallyExclusive: true,
        description: null,
        additionalAnswer: null,
        qCode: "20",
        type: "MutuallyExclusiveOption",
        option: true,
      },
    ];
```

Comment (tjamesmac) [inline eq-author/src/App/qcodes/QCodesTable/index.test.js:290]: The formatting is a bit weird here

Comment (tjamesmac) [reply eq-author/src/App/qcodes/QCodesTable/index.test.js:290]:
```
     {
        title: "<p>Questions 1</p>",
        alias: undefined,
        id: "ans-p1-1",
        description: "",
        guidance: "",
        label: "num1",
        qCode: "123",
        secondaryQCode: "1",
        type: "Number",
        questionPageId: "qp-1",
        secondaryLabel: null,
      }
```

File: eq-author/src/App/qcodes/QCodesTable/index.test.js [hunk 11]
```diff
@@ -34,305 +21,278 @@ const dummyQcodes = {
...
+
+  duplicates = {123: 2};
```
Comment (tjamesmac) [inline eq-author/src/App/qcodes/QCodesTable/index.test.js:292]:
```suggestion
  duplicates = { 123: 2 };
```

File: eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js [hunk 12]
```diff
@@ -50,12 +52,82 @@ describe("MainNavigation", () => {
...
+        label: 'Dur1',
+        qCode: 'qCode3',
+        secondaryQCode: '5',
+        type: 'Duration',
+        questionPageId: 'qp-1',
+        secondaryLabel: null 
+      },
+    ];
+    duplicateQCode = false;
+    // duplicates = {qCode3: 2};
```
Comment (tjamesmac) [inline eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js:122]:
```suggestion
```

Comment (tjamesmac) [reply eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js:122]: The `flattenedAnswer` format is also off here

File: eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js [hunk 13]
```diff
@@ -50,12 +52,82 @@ describe("MainNavigation", () => {
...
   });
 
   it("should enable all buttons if there are no errors on questionnaire", () => {
     const { getByTestId } = render(
       <MeContext.Provider value={{ me: user }}>
-        <UnwrappedMainNavigation {...props} />
+         <QCodeContext.Provider value={{flattenedAnswers, duplicateQCode}}>
```
Comment (tjamesmac) [inline eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js:128]:
```suggestion
         <QCodeContext.Provider value={{ flattenedAnswers, duplicateQCode }}>
```

File: eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js [hunk 14]
```diff
@@ -80,15 +152,16 @@ describe("MainNavigation", () => {
...
-    // expect(publishBtn).not.toBeDisabled();
   });
 
   it("should disable qcodes, publish and preview buttons if there are errors on questionnaire", async () => {
     props.questionnaire.totalErrorCount = 1;
 
     const { getByTestId } = render(
       <MeContext.Provider value={{ me: user }}>
-        <UnwrappedMainNavigation {...props} />
+        <QCodeContext.Provider value={{flattenedAnswers, duplicateQCode}}>
```
Comment (tjamesmac) [inline eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js:162]:
```suggestion
        <QCodeContext.Provider value={{ flattenedAnswers, duplicateQCode }}>
```

File: eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js [hunk 15]
```diff
@@ -109,14 +182,54 @@ describe("MainNavigation", () => {
...
     expect(qcodesBtn.hasAttribute("disabled")).toBeTruthy();
-    // expect(publishBtn.hasAttribute("disabled")).toBeTruthy();
+  });
+
+  it("should provide the validation error dot for the QCodes tab if there is an empty qCode", async () => {
+    flattenedAnswers[0].qCode = "";
+    
+    const { getByTestId } = render(
+      <MeContext.Provider value={{ me: user }}>
+        <QCodeContext.Provider value={{flattenedAnswers, duplicateQCode}}>
```
Comment (tjamesmac) [inline eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js:199]:
```suggestion
        <QCodeContext.Provider value={{ flattenedAnswers, duplicateQCode }}>
```

File: eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js [hunk 16]
```diff
@@ -109,14 +182,54 @@ describe("MainNavigation", () => {
...
+
+    expect(getByTestId("small-badge")).toBeTruthy();
+  });
+
+  it("should provide the validation error dot for the QCodes tab if there are duplicate qCodes", async () => {
+    duplicateQCode = true;
+
+    const { getByTestId } = render(
+      <MeContext.Provider value={{ me: user }}>
+        <QCodeContext.Provider value={{flattenedAnswers, duplicateQCode}}>
```
Comment (tjamesmac) [inline eq-author/src/App/QuestionnaireDesignPage/MainNavigation/index.test.js:220]:
```suggestion
        <QCodeContext.Provider value={{ flattenedAnswers, duplicateQCode }}>
```

**Review by tjamesmac — DISMISSED** [review]
Looks really good Shane, nice one! Just a few formatting issues here and there. Think I highlighted most of them

**Review by farres1 — APPROVED** [review]

**Review by tbaggaley — APPROVED** [review]

**Pull request merged**